runtime/src/qio/qio.c

```diff
@@ -2798,56 +2802,140 @@ qioerr _qio_buffered_read(qio_channel_t* ch, void* ptr, ssize_t len, ssize_t* am
+      start = _right_mark_start_iter(ch);            // start of available data
       end = start;
-      qbuffer_iter_advance(&ch->buf, &end, gotlen);
-    } else {
-      end = _av_end_iter(ch);
+      qbuffer_iter_advance(&ch->buf, &end, gotlen);  // end of available data
+
+      // copy 'gotlen' bytes into the ptr
+      err = qbuffer_copyout(&ch->buf, start, end, ptr, gotlen);
+      if( err ) return err;
+
+      // advance the ptr, start of aviliabe data, etc.
+      ptr = qio_ptr_add(ptr, gotlen);
+      _set_right_mark_start(ch, end.offset);
+      remaining -= gotlen;
+      *amt_read = gotlen;
+
+      // clean up the now unused portion of the buffer
+      qbuffer_trim_front(&ch->buf, gotlen);
     }
 
-    // Now copy out the data.
-    err = qbuffer_copyout(&ch->buf, start, end, (char *) ptr + (len-remaining),
-                          gotlen);
-    if( err ) goto error;
+    // make a direct system call to read the rest
+    while( remaining > 0 ) {
+      num_read = 0;
+      switch (method) {
+        case QIO_METHOD_READWRITE:
+          err = qio_int_to_err(sys_read(ch->file->fd, ptr, remaining,
+                               &num_read));
+          break;
+        case QIO_METHOD_PREADPWRITE:
+          err = qio_int_to_err(sys_pread(ch->file->fd, ptr, remaining,
+                               _right_mark_start(ch), &num_read));
+          break;
+        case QIO_METHOD_FREADFWRITE:
+          if( ch->file->fp ) {
+            num_read_u = fread(ptr, 1, remaining, ch->file->fp);
+            err = 0;
+            if( num_read_u == 0 ) {
+              if( feof(ch->file->fp) ) err = QIO_EEOF;
+              else err = qio_int_to_err(ferror(ch->file->fp));
+            }
+            num_read = num_read_u;
+          } else {
+            QIO_GET_CONSTANT_ERROR(err, EINVAL, "missing file pointer");
+            num_read = 0;
+          }
+          break;
+        case QIO_METHOD_MMAP:
+          break;
+        case QIO_METHOD_MEMORY:
+          break;
+      }
+      // Return early on an error or on EOF.
+      if( err ) {
+        *amt_read = len - remaining;
+        return err;
+      }
+      ptr = qio_ptr_add(ptr, num_read);
+      _add_right_mark_start(ch, num_read);
+      remaining -= num_read;
+      *amt_read += num_read;
+    }
 
-    // now advance the start of the available buffer by the amount.
-    _set_right_mark_start(ch, end.offset);
+    // reposition the existing buffer space (and 'av_end') by the size of the direct read
+    // this way, the buffer is ready for any future buffered reads
+    // (the buffer is empty, so we can just move the end pointers)
+    qbuffer_reposition(&ch->buf, qbuffer_start_offset(&ch->buf) +
+                       *amt_read - gotlen);
+    ch->av_end += *amt_read - gotlen;
 
-    // move the iterator for the next read
-    qbuffer_iter_advance(&ch->buf, &start, gotlen);
+    // re-setup the buffer for any future buffered writes
+    _qio_buffered_setup_cached(ch);
```

Comment:
Now that you have some performance testing, you can answer the question, is it faster if you leave this line out?